picamera2/picamera2.py

```diff
@@ -152,7 +152,7 @@ def _reset_flags(self) -> None:
         self.pre_callback = None
         self.post_callback = None
         self.completed_requests = []
-        self.lock = threading.Lock()  # protects the functions and completed_requests fields
+        self.lock = threading.RLock()  # protects the functions and completed_requests fields
```

Comment:
Use of a reentrant lock obviates the need for the protected vs. unprotected versions of `dispatch_functions()`